### PR TITLE
Local Econet and Level-3 File Server support

### PR DESCRIPTION
**Pull request by robsonjb1**
Hi Matt,
I've ported over the Econet emulation from Beebem together with the FSEM2 File server emulator as a kind of hybrid mash-up. It was done for fun really but thought you may wish to include it in core jsbeeb?

There is a new 'Econet' checkbox against the machine configuration. The BBC 1770 DFS and any Master can be used (DFS 0.9 does not include NFS). Master series machines automatically include the ANFS filing system.

The file server uses a 10mb drive image (I've included the one that comes with FSEM2). If Econet is included, a new menu item appears under 'Disks' allowing upload/download of the file server VM's.
To use, ensure the NFS filing system is selected (*NET) and log in with the standard *I AM SYST. The BBC Welcome disk is included on the default file server image. Log in with the user *I AM WELCOME to auto-boot into this.

It should be noted that this is a LOCAL ONLY emulation of Econet - no network traffic goes outside the browser. Use this to play around with an L3 Econet file server and the BBC and Master network filing systems.

I am hosting this branch here if you wish to take a look:
https://jrbeeb.azurewebsites.net

### CODE REVIEW

File: 6502.js [hunk 1]
```diff
@@ -601,6 +603,24 @@ export function Cpu6502(model, dbgr, video_, soundChip_, ddNoise_, music5000_, c
             }
         }
 
+        if (this.econet) {
```
Comment (mattgodbolt) [inline 6502.js:606]: I might see if I can rephrase this in the switch() statement below: ideally that should be the spot where most of the work is done (else we'll end up with more and more stuff here being checked before every hw access, which seems a shame.

File: 6502.js [hunk 2]
```diff
@@ -908,7 +944,12 @@ export function Cpu6502(model, dbgr, video_, soundChip_, ddNoise_, music5000_, c
             const awaiting = [];
 
             for (let i = 0; i < extraRoms.length; ++i) {
+                // Skip over banks 4-7 (sideways RAM on a Master)
                 romIndex--;
+                while (capturedThis.model.swram[romIndex]) {
```
Comment (mattgodbolt) [inline 6502.js:949]: Oh nice! Thanks for fixing this!

File: 6502.js [hunk 3]
```diff
@@ -1035,6 +1081,13 @@ export function Cpu6502(model, dbgr, video_, soundChip_, ddNoise_, music5000_, c
         this.tube.execute(cycles);
         if (this.teletextAdaptor) this.teletextAdaptor.polltime(cycles);
         if (this.music5000) this.music5000.polltime(cycles);
+        if (this.econet) {
+            let donmi = this.econet.polltime(cycles);
```
Comment (mattgodbolt) [inline 6502.js:1085]: this can be `const`

File: config.js [hunk 4]
```diff
@@ -32,6 +33,16 @@ export function Config(onClose) {
         this.model.tube = enabled ? findModel("Tube65c02") : null;
     };
 
+    this.setEconet = function (enabled) {
+        enabled = !!enabled;
+        $("#hasEconet").prop("checked", enabled);
+        this.model.hasEconet = enabled;
+
+        if (enabled && this.model.name.toString().includes("Master")) {
```
Comment (mattgodbolt) [inline config.js:41]: I'd rather make "is it a master" more explicit than the string check here: though this is OK for now

File: filestore.js [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+                console.log("Filestore: unhandled osbyte 0x" + this.A.toString(16));
+                break;
+        }
+    }
+
+    reset() {
+        console.log("Filestore: initialisation");
+
+        const filestoreRef = this;
+        utils.loadData("econet/L3FS.dat").then(function (data) {
```
Comment (mattgodbolt) [inline filestore.js:313]: Wow, so much to learn about how this worked!

**Review by mattgodbolt — APPROVED** [review]
All looks so good! I'm going to merge and tackle the (super minor!) things myself. Amazing :) Thank you so much!

**Comment by mattgodbolt** [comment]
I've toyed with the idea of a real remote econet, fwiw... I can probably stand up a fake endpoint pretty easily :) But...this is grea. Thank you once again!

**Comment by mattgodbolt** [comment]
That said: I merged and it seemed to wedge: I selected at "BBC Master 128 (ANFS)" and then ticked "Econet" ... `*.` did nothing and now ctrl-BREAK is:

image

... in the JS output I see:

```

filestore.js:61 Filestore: Acorn File Server Level 3 Version 1.26
filestore.js:61 Filestore: Version string   File server station     254    '
filestore.js:61 Filestore: Testing Memory
filestore.js:61 Filestore: 
filestore.js:61 Filestore: Last free location = F7FF
filestore.js:61 Filestore: Number of drives:1
filestore.js:61 Filestore: Command :S
filestore.js:61 Filestore: Stations:1
filestore.js:61 Filestore: Cache size - 45A3 objects - 45E
filestore.js:61 Filestore: Starting - Ready
```

but it's hung completely. Any ideas what I'm doing wrong?

File: config.js (hunk 4 above)
Comment (mattgodbolt) [reply config.js:41]: Oh! There's already an `isMaster` :) will change

**Comment by mattgodbolt** [comment]
(eventually I got "Line Jammed")

**Comment by mattgodbolt** [comment]
I committed my changes: I had seen the issues prior to my tweaks but it's not impossible that I broke something! So please if you could take a look that would be great.

**Comment by robsonjb1** [comment]
> I committed my changes: I had seen the issues prior to my tweaks but it's not impossible that I broke something! So please if you could take a look that would be great.

Hi Matt - I think I know what this could be. If you have already used jsbeeb before as a Master then you'll have a persisted CMOS store that *doesn't* include the defaults I added for the FS station number. So in this case it would try to contact a FS at station 0 which is invalid. This gives a Line Jammed with a *. (although I can't replicate a freeze on a CTRL+BREAK).

I'll sort this out tonight after work (UK time).
Thanks,
Jason.

**Comment by robsonjb1** [comment]
Thanks Matt!

On 28 Mar 2023, at 18:04, Matt Godbolt ***@***.***> wrote:



Nice find and fix! Should be deploying now :)

**Pull request merged**